system/lib/pthread/library_pthread.c

```diff
@@ -291,54 +309,119 @@ static void _do_call(em_queued_call *q)
+	} else {
+		// The caller owns this call object, it is listening to it and will free it up.
+		q->operationDone = 1;
+		emscripten_futex_wake(&q->operationDone, INT_MAX);
+	}
 }
 
 #define CALL_QUEUE_SIZE 128
 static em_queued_call **call_queue = 0;
-static int call_queue_length = 0; // Shared data synchronized by call_queue_lock.
+static volatile int call_queue_head = 0; // Shared data synchronized by call_queue_lock.
+static volatile int call_queue_tail = 0;
 static pthread_mutex_t call_queue_lock = PTHREAD_MUTEX_INITIALIZER;
 
-void EMSCRIPTEN_KEEPALIVE emscripten_sync_run_in_main_thread(em_queued_call *call)
+EMSCRIPTEN_RESULT emscripten_wait_for_call_v(em_queued_call *call, double timeoutMSecs)
+{
+	int r;
+
+	int done = emscripten_atomic_load_u32(&call->operationDone);
+	if (!done) {
+		double now = emscripten_get_now();
+		double waitEndTime = now + timeoutMSecs;
+		emscripten_set_current_thread_status(EM_THREAD_STATUS_WAITPROXY);
+		while(!done && now < waitEndTime) {
+			r = emscripten_futex_wait(&call->operationDone, 0, waitEndTime - now);
+			done = emscripten_atomic_load_u32(&call->operationDone);
+			now = emscripten_get_now();
+		}
+		emscripten_set_current_thread_status(EM_THREAD_STATUS_RUNNING);
+	}
+	if (done) return EMSCRIPTEN_RESULT_SUCCESS;
+	else return EMSCRIPTEN_RESULT_TIMED_OUT;
+}
+
+EMSCRIPTEN_RESULT emscripten_wait_for_call_i(em_queued_call *call, double timeoutMSecs, int *outResult)
+{
+	EMSCRIPTEN_RESULT res = emscripten_wait_for_call_v(call, timeoutMSecs);
+	if (res == EMSCRIPTEN_RESULT_SUCCESS && outResult) *outResult = call->returnValue.i;
+	return res;
+}
+
+void EMSCRIPTEN_KEEPALIVE emscripten_async_run_in_main_thread(em_queued_call *call)
 {
 	assert(call);
+	// If we are the main Emscripten runtime thread, we can just call the operation directly.
 	if (emscripten_is_main_runtime_thread()) {
 		_do_call(call);
 		return;
 	}
+
+	// Add the operation to the call queue of the main runtime thread.
 	pthread_mutex_lock(&call_queue_lock);
 	if (!call_queue) call_queue = malloc(sizeof(em_queued_call*) * CALL_QUEUE_SIZE); // Shared data synchronized by call_queue_lock.
-	// Note: currently call_queue_length can be at most the number of pthreads that are currently running, so the queue can never get
-	// full. However if/when the queue is extended to be asynchronous for void-returning functions later, this will need to be revised.
-	assert(call_queue_length < CALL_QUEUE_SIZE);
-	call_queue[call_queue_length] = call;
-	++call_queue_length;
-	if (call_queue_length == 1) {
+
+	int head = emscripten_atomic_load_u32((void*)&call_queue_head);
+	int tail = emscripten_atomic_load_u32((void*)&call_queue_tail);
+	int new_tail = (tail + 1) % CALL_QUEUE_SIZE;
+
+	while(new_tail == head) { // Queue is full?
+		pthread_mutex_unlock(&call_queue_lock);
+		emscripten_futex_wait((void*)&call_queue_head, head, INFINITY);
+		pthread_mutex_lock(&call_queue_lock);
+		head = emscripten_atomic_load_u32((void*)&call_queue_head);
```

Comment:
That is correct, however `volatile` is still needed, which in Emscripten compiles down to `emscripten_atomic_load_u32`, (because there is no concept of volatile correctness in JavaScript engines), so this line is the same thing. Having an unqualified line like `head = call_queue_head;` to read a volatile variable would look odd, so prefer explicitly qualifying instead.